src/order/filter/bornology.lean

```diff
@@ -0,0 +1,138 @@
+    generalize : sᶜ = s',
+    refine λ h, h.dinduction_on _ (λ x t hx ht h, _),
+    { exact empty_mem, },
+    { refine insert_eq x t ▸ union_mem _ _ _ h,
+      obtain ⟨b, hb : b ∈ B, hxb : x ∈ b⟩ :=
+        mem_sUnion.mp (by simpa [←sUnion_univ] using mem_univ x),
+      exact subset_mem _ hb _ (singleton_subset_iff.mpr hxb) },
+  end }
+
+section bornological_space
+
+@[ext]
+lemma bornological_space.eq : ∀ {f g : bornological_space α}, f.cobounded = g.cobounded → f = g
+| ⟨a, _⟩ ⟨b, _⟩ rfl := rfl
+
+section
+variables [t : bornological_space α] {s₁ s₂ : set α}
+include t
+
+/-- `cobounded` is the filter of complements of bounded sets in the ambient bornological
+space on `α` -/
+def cobounded : filter α := bornological_space.cobounded t
+
+/-- `is_bounded` is the predicate that `s` is bounded relative to the ambient bornological
+space on `α`. -/
+def is_bounded (s : set α) : Prop := sᶜ ∈ (cobounded : filter α)
+
+@[simp] lemma le_cofinite : cobounded ≤ (cofinite : filter α) :=
+bornological_space.le_cofinite t
+
+lemma is_bounded_def {s : set α} : is_bounded s ↔ sᶜ ∈ t.cobounded := iff.rfl
+
+lemma is_bounded_compl_iff {s : set α} : is_bounded sᶜ ↔ s ∈ t.cobounded :=
+by rw [is_bounded_def, compl_compl]
+
+@[simp]
+lemma is_bounded_empty : is_bounded (∅ : set α) :=
+by { rw [is_bounded_def, compl_empty], exact univ_mem}
+
+lemma is_bounded.union (h₁ : is_bounded s₁) (h₂ : is_bounded s₂) : is_bounded (s₁ ∪ s₂) :=
+by { rw [is_bounded_def, compl_union], exact t.cobounded.inter_sets h₁ h₂ }
+
+lemma is_bounded.subset (h₁ : is_bounded s₂) (h₂ : s₁ ⊆ s₂) : is_bounded s₁ :=
+by { rw [is_bounded_def], exact t.cobounded.sets_of_superset h₁ (compl_subset_compl.mpr h₂) }
+
+@[simp]
+lemma sUnion_bounded_univ : (⋃₀ {s : set α | is_bounded s}) = set.univ :=
+univ_subset_iff.mp $ λ x hx, mem_sUnion_of_mem (mem_singleton x)
+  $ le_def.mp t.le_cofinite {x}ᶜ $ (set.finite_singleton x).compl_mem_cofinite
+
+end
+
+end bornological_space
+
+namespace bornological_space
+
+lemma eq_iff {t t' : bornological_space α} :
+  t = t' ↔ ∀ s, (@cobounded α t).sets s ↔ (@cobounded α t').sets s :=
+⟨λ h s, h ▸ iff.rfl, λ h, by { ext, exact h _ }⟩
+
+lemma eq_iff_is_bounded {t t' : bornological_space α} :
+  t = t' ↔ ∀ s, @is_bounded α t s ↔ @is_bounded α t' s :=
+⟨λ h s, h ▸ iff.rfl, λ h, by { ext, simpa only [is_bounded_def, compl_compl] using h sᶜ, }⟩
+
+variables [bornological_space α] [bornological_space β] [bornological_space γ]
+
+lemma is_bounded_sUnion {s : set (set α)} (hs : finite s) :
+  (∀ t ∈ s, is_bounded t) → is_bounded (⋃₀ s) :=
+finite.induction_on hs (λ _, by rw sUnion_empty; exact is_bounded_empty) $
+λ a s has hs ih h, by rw sUnion_insert; exact
+is_bounded.union (h _ $ mem_insert _ _) (ih $ λ t, h t ∘ mem_insert_of_mem _)
+
+lemma is_bounded_bUnion {s : set β} {f : β → set α} (hs : finite s) :
+  (∀ i ∈ s, is_bounded (f i)) → is_bounded (⋃ i ∈ s, f i) :=
+finite.induction_on hs
+  (λ _, by rw bUnion_empty; exact is_bounded_empty)
```

Comment:
Here and elsewhere: please only use `;` when the first tactic actually produces more than one goal.
```suggestion
  (λ _, by { rw bUnion_empty, exact is_bounded_empty })
```

Reply:
Sorry, these bits were copy-pasted from `topology/basic`, but I have fixed the usage of `;` here.